src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,252 @@
+static bool isReplicationOriginCreated(MultiConnection *connection, char *originName,
+									   RepOriginId *originId);
+static RepOriginId CreateReplicationOriginSession(MultiConnection *connection,
+												  char *originName);
+static void ReplicationOriginSessionSetupHelper(MultiConnection *connection,
+												char *originName);
+static bool ExecuteRemoteCommandAndCheckResult(MultiConnection *connection, char *command,
+											   char *expected);
+
+/* SetupReplicationOriginRemoteSession sets up a new replication origin session in a
+ * remote session. The identifier is used to create a unique replication origin name
+ * for the session in the remote node.
+ */
+void
+SetupReplicationOriginRemoteSession(MultiConnection *connection, char *identifier)
+{
+	if (!isReplicationOriginSessionSetup(connection))
+	{
+		StringInfo originNameString = makeStringInfo();
+		appendStringInfo(originNameString, "citus_internal_%d_%s", GetLocalNodeId(),
+						 identifier);
+		char *originName = quote_literal_cstr(originNameString->data);
+		ReplicationOriginSessionSetupHelper(connection, originName);
+	}
+}
+
+
+/* ResetReplicationOriginRemoteSession resets the replication origin session in a
+ * remote node.
+ */
+void
+ResetReplicationOriginRemoteSession(MultiConnection *connection, char *identifier)
+{
+	/* Reset the replication origin in remote session*/
+	StringInfo replicationOriginSessionResetQuery = makeStringInfo();
+	appendStringInfo(replicationOriginSessionResetQuery,
+					 "select pg_catalog.pg_replication_origin_session_reset();");
+	ExecuteCriticalRemoteCommand(connection,
+								 replicationOriginSessionResetQuery->data);
+
+	/* Drop the replication origin entry created in remote session.*/
+	StringInfo originNameString = makeStringInfo();
+	appendStringInfo(originNameString, "citus_internal_%d_%s", GetLocalNodeId(),
+					 identifier);
+	StringInfo replicationOriginSessionDropQuery = makeStringInfo();
+	appendStringInfo(replicationOriginSessionDropQuery,
+					 "select pg_catalog.pg_replication_origin_drop(%s);",
+					 quote_literal_cstr(originNameString->data));
+
+	ExecuteCriticalRemoteCommand(connection,
+								 replicationOriginSessionDropQuery->data);
+}
+
+
+/* SetupReplicationOriginLocalSession sets up a new replication origin session in a
+ * local session.
+ */
+void
+SetupReplicationOriginLocalSession(void)
+{
+	if (!isReplicationOriginSessionSetup(NULL))
+	{
+		StringInfo originNameString = makeStringInfo();
+		appendStringInfo(originNameString, "citus_internal_%d", GetLocalNodeId());
+		ReplicationOriginSessionSetupHelper(NULL, originNameString->data);
+	}
+}
+
+
+/* ResetReplicationOriginLocalSession resets the replication origin session in a
+ * local node.
+ */
+void
+ResetReplicationOriginLocalSession(void)
+{
+	/*Reset Replication Origin in local session */
+	if (replorigin_session_origin != InvalidRepOriginId)
+	{
+		replorigin_session_reset();
+		replorigin_session_origin = InvalidRepOriginId;
```

Comment:
should we consider a case where the client had set an origin ID?

Reply:
yes, possible, but I am not sure how to remember that.. 
Should I intoruduce a ReplicationOriginContext and remember this in that 
and return that pointer to the caller? 
Then during Reset operation, the caller can give that pointer back, and the reset function can restore it back..
It is possible, just adds a bit more complexity..